fix: auto-compact on context limit error

kwsantiago opened the pull request:
## Summary
- When context limit is exceeded, automatically compact the conversation to continue instead of failing.
- Fixes https://github.com/block/goose/issues/1303

## Changes
- Catch `ProviderError::ContextLengthExceeded` in agent reply loop
- Call `auto_compact::perform_compaction` to reduce conversation size without threshold checking
- Refactored compaction logic to separate threshold checking from actual compaction
- Continue conversation if compaction succeeds, fail gracefully with clear error message if not

## Testing
- [X] Builds successfully
- [X] Existing tests pass
- [X] Clippy passes with no warnings
- [X] Code formatted with cargo fmt

michaelneale commented on crates/goose/src/providers/factory.rs:
```diff
@@ -162,7 +193,20 @@ fn create_provider(name: &str, model: ModelConfig) -> Result<Arc<dyn Provider>>
+        "openrouter" => {
+            let mut provider = OpenRouterProvider::from_env(model)?;
+
+            // Check if middle-out should be enabled by default
```
ideally could this be in the from_env implementation encapsulated there?

michaelneale commented on crates/goose/src/providers/utils.rs:
```diff
@@ -347,6 +347,63 @@ pub fn emit_debug_trace<T1, T2>(
 
+/// Parse token information from common error message formats
+pub fn parse_token_info_from_error(error_msg: &str) -> (Option<u32>, Option<u32>) {
+    // OpenRouter format: "This endpoint's maximum context length is 200000 tokens. However, you requested about 201326 tokens"
```
is there similar for openai and other big ones?

michaelneale commented:
thanks @kwsantiago this looks interesting, seems a few things combined (wasn't aware of that openrouter stuf f- we have some other compaction things happening so hopefully wont' hit that or need it?). 

Can you talk a bit more about how this prevents - there is new state with the token counting, not sure if that helps with it, or is it sniffing out error messages. I wonder if the openrouter changex can be a separate PR to make this change clearer, but I like the direction

michaelneale commented on crates/goose/src/agents/agent.rs:
```diff
@@ -220,6 +234,21 @@ impl Agent {
+    }
+
+    /// Enable middle-out transform for OpenRouter provider
+    pub fn enable_middle_out_transform(&self) {
```
I don't think this should be in the agent, is too specific to openrouter at the moment

michaelneale commented on crates/goose/src/agents/agent.rs:
```diff
@@ -152,6 +155,7 @@ impl Agent {
             router_tool_selector: Mutex::new(None),
             scheduler_service: Mutex::new(None),
             retry_manager,
+            token_tracker: create_shared_tracker(),
```
slightly worried about keeping state here - not sure why but something to think about

kwsantiago commented:
@michaelneale Thank you for the feedback!

The current state of this PR is that it parses error messages after hitting the limit and stores the discovered limit in a token tracker, but doesn't use this to prevent future errors.

What I think we can do here is split this out into 2 PRs:
- This PR (core fix):
  - Keep the error parsing and improved error messaging
  - Add retry logic that automatically enables middle-out for
   OpenRouter after detecting a context limit error
  - Remove the token tracker state (since it isn't being used for
  prevention)

- Separate PR (OpenRouter enhancements):
  - Move OpenRouter-specific middle-out configuration to the
  provider's `from_env` method
  - Add `OPENROUTER_ENABLE_MIDDLE_OUT` env var support
  - Implement proactive token counting if needed

I'll begin working on this, but if you have any other feedback, please let me know.

kwsantiago commented on crates/goose-cli/src/session/mod.rs:
```diff
@@ -1143,6 +1143,32 @@ impl Session {
+                                            None
+                                        )
+                                        .await?;
+                                    continue;
```
Thank you for catching! I added a commit to fix.

kwsantiago commented:
@DOsinga let me know what you think. Based on your [proposed solution from the Discord thread](https://discord.com/channels/1287729918100246654/1399727098574143508/1399775229353529506), I added an initial implementation for automatic recovery when tool responses are too big and hit context limits. In this way, Goose recovers automatically instead of dying silently when tools return huge responses.

DOsinga commented:
I think @katzdave should have a look how this fits in with his efforts

DOsinga commented:
Ok, so a previous attempt to unify http clients was this:
https://github.com/block/goose/pull/3547

I'm currently working on this:
https://github.com/block/goose/pull/3558

to unify some other stuff, but was about to copy the client sharing from the firrst (closed) PR. I'm thinking we should get that in and that should make your life easier. I think having a fixed timeout for now would be fine actually for now. I imagine we replace all the the environment variables with true settings soon anyway.

_ceterum censeo variabiles ambientis esse delendas_

michaelneale commented:
@kwsantiago very nice looking focussed - some changes merged recently do you want to update this to main? (I could also push to it if you like). Curious if this helps with the openrouter context exhaustion

kwsantiago commented:
@michaelneale sounds good, if you don't get to it, I can do this tonight (EST).

Regarding open router middle-out support, should I resurrect that in a follow up PR or within this PR? I know we had originally discussed taking it out which is why I got rid of it in this branch, although it is in some of my earlier commits if we want to dig it back out.

michaelneale commented:
@kwsantiago yeah if you want to add a middle-out in another PR - go ahead (I have an open PR which enables some defaults for startup experience so could always do it there) - if you have time, go ahead and tag me

michaelneale commented on crates/goose-cli/src/session/mod.rs:
```diff
@@ -1218,6 +1218,83 @@ impl Session {
                         }
 
                         Some(Err(e)) => {
+                            // Check if this is a context length error
```
oh - @kwsantiago I think this needs to be in the goose crate - as the GUI will hit this as well won't it? ie not just the cli

kwsantiago replied:
@michaelneale as in, move it out completely from CLI->GUI, or simply replicate this in the GUI as well?

michaelneale replied:
if possible - would be ideal in main goose crate, just once, but if not possible, goose-server will need it replicated, but that isn't ideal as it seems session specific code, not client code

michaelneale reviewed (changes requested):
hoping we can move it to the session stuff in the core goose crate not CLI

kwsantiago commented:
@michaelneale I moved the detection logic to `agents/agent.rs` in my last commit.

Let me know if we need to adjust anything there.

michaelneale commented on crates/goose-mcp/src/developer/mod.rs:
```diff
@@ -2329,33 +2329,28 @@ mod tests {
             .as_ref()
-            .map_or(false, |desc| desc
-                .contains("Replace a string in a file with a new string")));
+            .is_some_and(|desc| desc.contains("Replace a string in a file with a new string")));
```
this is just a style/non functional change?

kwsantiago replied:
@michaelneale This was from clippy, completely unrelated to this PR, but figured I'd add it in since CI would probably fail there.

Can remove if needed.

michaelneale commented on crates/goose-cli/src/session/mod.rs:
```diff
@@ -1226,6 +1226,28 @@ impl Session {
 
                         Some(Err(e)) => {
+                            // Check if this is a context length error that wasn't already handled by the agent
+                            if let Some(goose::providers::errors::ProviderError::ContextLengthExceeded(error_msg)) = e.downcast_ref::<goose::providers::errors::ProviderError>() {
```
not sure this belongs here - if so, how do we show it in GUI? I think could probably do without this, the agent is doing the heavy lifting and I dont' think we want any special provider code in here either - and won't need to say middle-out

kwsantiago replied:
Makes sense, I'll update to reflect.

michaelneale commented:
looking good @kwsantiago - are you happy with the tests in the cli module - I can't quite tell if they are testing things there or really are testing things which are (now correctly) in agent module?

michaelneale commented on crates/goose/src/agents/agent.rs:
```diff
@@ -1083,11 +1083,60 @@ impl Agent {
+                            } else {
+                                // Not caused by tool response, show the error to user
+                                yield AgentEvent::Message(Message::assistant().with_context_length_exceeded(
+                                    format!("Context length exceeded: {}", error_msg)
```
can we add in "please try a new session" like before @kwsantiago ?

michaelneale replied:
ie similar message to before

michaelneale commented:
I think looks ok @katzdave would like your 👍 to merge this (maybe could tweak the fallback error message) to make sure it doesn't flag anything for you

michaelneale commented:
@katzdave ok with me - if you want to give a 👍

kwsantiago commented:
Not that I'm aware of @katzdave , unless you see anything here that needs to be changed.

katzdave commented:
I think we can probably close this.  https://github.com/block/goose/pull/4202 addresses this out of context issue.

katzdave reviewed (commented):
Gave this some testing on cli:

Seems like I hit the context error when the agent pulls my last file read:

```
Error: Request failed: HTTP 400 Bad Request: {"error_code":"BAD_REQUEST","message":"{\"external_model_provider\":\"anthropic\",\"external_model_error\":{\"type\":\"error\",\"error\":{\"type\":\"invalid_request_error\",\"message\":\"input length and `max_tokens` exceed context limit: 196758 + 8192 > 200000, decrease input length or `max_tokens` and try again\"},\"request_id\":\"req_011CSjw2Yhikd24PTTPDVY6s\"}}"}
The tool calling loop was interrupted. How would you like to proceed?
  error: The error above was an exception we were not able to handle.
These errors are often related to connection or authentication
We've removed the conversation up to the most recent user message
- depending on the error you may be able to continue
```

on desktop:

```
13:04:26.166 › goosed stderr for port 49736 and dir /Users/dkatz/git/goose3/goose:   2025-09-02T17:04:26.166432Z  INFO goosed::routes::reply: Session started, counter.goose.session_starts: 1, session_type: "app", interface: "ui"
    at crates/goose-server/src/routes/reply.rs:179


13:04:26.213 › from renderer: Displaying content for message: encjh7cz content: hi keep looking around
13:04:26.213 › from renderer: Displaying content for message: encjh7cz content: hi keep looking around
13:04:27.840 › goosed stderr for port 49736 and dir /Users/dkatz/git/goose3/goose:   2025-09-02T17:04:27.839627Z ERROR goosed::routes::reply: Error processing message: Request failed: HTTP 400 Bad Request: {"error_code":"BAD_REQUEST","message":"{\"external_model_provider\":\"anthropic\",\"external_model_error\":{\"type\":\"error\",\"error\":{\"type\":\"invalid_request_error\",\"message\":\"input length and `max_tokens` exceed context limit: 196869 + 8192 > 200000, decrease input length or `max_tokens` and try again\"},\"request_id\":\"req_011CSjwJNaxUPqb88BtnhU34\"}}"}
    at crates/goose-server/src/routes/reply.rs:303


13:04:27.842 › goosed stderr for port 49736 and dir /Users/dkatz/git/goose3/goose:   2025-09-02T17:04:27.840557Z  INFO goosed::routes::reply: Session completed, counter.goose.session_completions: 1, session_type: "app", interface: "ui", exit_type: "normal", duration_ms: 1673, total_tokens: 0, message_count: 385
    at crates/goose-server/src/routes/reply.rs:353

  2025-09-02T17:04:27.840686Z  INFO goosed::routes::reply: Session duration, counter.goose.session_duration_ms: 1673, session_type: "app", interface: "ui"
    at crates/goose-server/src/routes/reply.rs:364
```

But doesn't actually trigger this new forced compaction flow. Was using databricks provider for this testing, and I believe that message does pass the check to trigger contextLengthExceeded error.

Is this working for you on a different provider?

kwsantiago commented:
@katzdave Thanks for the detailed feedback! I've addressed the issues:

1. Reverted to using true context limit instead of `estimate_target_context_limit` as requested
2. Added fallback detection for context errors in message text to handle databricks' nested error format
3. The fix should now catch errors like "input length and max_tokens exceed context limit" and trigger auto-compaction

I don't have access to test with actual providers (databricks/openrouter), so I'd appreciate if someone could verify the fix works end-to-end. I won't have much bandwidth to iterate much further on this, but the changes should handle the reported issues. Feel free to edit the branch as needed.

michaelneale commented:
@katzdave both @zanesq and I see that databricks error too, unrelated to this change, so it may be an unrelated regression?

michaelneale commented:
@kwsantiago can you run fmt again?

kwsantiago commented:
@michaelneale just commited.

michaelneale commented:
@kwsantiago sorry again - but can you update to main, hopefully will be last time. Will try this now, I think this may be good.

michaelneale commented on crates/goose/src/providers/utils.rs:
```diff
@@ -163,6 +163,18 @@ pub async fn handle_status_openai_compat(response: Response) -> Result<Response,
+        let body_lower = body_str.to_lowercase();
+        if body_lower.contains("context limit")
+            || body_lower.contains("context_length_exceeded")
+            || (body_lower.contains("input length") && body_lower.contains("exceed"))
```
wonder if could add `Input is too long` or even just `Input` - and ignore case? (I might try this)

kwsantiago replied:
@michaelneale I pushed up a commit for this, try it out.

michaelneale reviewed (approved):
nice @kwsantiago - took a while but got there in the end!

katzdave reviewed (commented):
Yes, nice work! I think we can revert the changes in errors.rs completely; taking it for one last spin quickly will update shortly.

kwsantiago commented:
@michaelneale that was due to addressing @katzdave requested changes here:

- https://github.com/block/goose/pull/3635#pullrequestreview-3183001543
- https://github.com/block/goose/pull/3635#discussion_r2320550389
- https://github.com/block/goose/pull/3635#discussion_r2320552269
- https://github.com/block/goose/pull/3635#discussion_r2320554968
- https://github.com/block/goose/pull/3635#issuecomment-3251302797

michaelneale reviewed (approved):
nice, thanks! took it for quick spin
